### PR TITLE
Enable fp8e5m2fnuz type

### PR DESCRIPTION
**Pull request by CharlieL7**
* Enables the E5M2 FNUZ datatype so that we have full support for all the current FP8 types
* E5M2 FNUZ will not be used directly in model but might be useful when converting from OCP -> FNUZ
* Lots of updated files, but mostly expanding test cases

### CODE REVIEW

File: src/api/include/migraphx/migraphx.h [hunk 1]
```diff
@@ -46,6 +46,7 @@
     m(uint32_type, uint32_t) \
     m(uint64_type, uint64_t) \
     m(fp8e4m3fnuz_type, migraphx::fp8::fp8e4m3fnuz) \
+    m(fp8e5m2fnuz_type, migraphx::fp8::fp8e5m2fnuz) \
```
[the remark this answers is not in the document]
Comment (CharlieL7) [reply src/api/include/migraphx/migraphx.h]: I changed the order

File: test/verify/test_gemm_add_broadcast2.cpp [hunk 2]
```diff
@@ -55,5 +55,6 @@ template struct test_gemm_add_broadcast2<migraphx::shape::float_type>;
 // template struct test_gemm_add_broadcast2<migraphx::shape::half_type>; // fails with CK,
```
Comment (TedThemistokleous) [inline test/verify/test_gemm_add_broadcast2.cpp:55]: Do we have a CK feature for this or is there a workaround?

File: test/verify/test_gemm_add.cpp [hunk 3]
```diff
@@ -58,3 +58,6 @@ struct test_gemm_add : verify_program<test_gemm_add<DType>>
 template struct test_gemm_add<migraphx::shape::float_type>;
 template struct test_gemm_add<migraphx::shape::half_type>;
 // TODO template struct test_gemm_add<migraphx::shape::fp8e4m3fnuz_type>;
+// TODO template struct test_gemm_add<migraphx::shape::fp8e5m2fnuz_type>;
+// TODO template struct test_gemm_add<migraphx::shape::fp8e4m2fn_type>;
+// TODO template struct test_gemm_add<migraphx::shape::fp8e5m2_type>;
```
Comment (TedThemistokleous) [inline test/verify/test_gemm_add.cpp:63]: Why are these TODO?

File: test/verify/test_gemm_transposea.cpp [hunk 4]
```diff
@@ -46,6 +46,7 @@ struct test_gemm_transposea : verify_program<test_gemm_transposea<DType>>
 template struct test_gemm_transposea<migraphx::shape::float_type>;
 template struct test_gemm_transposea<migraphx::shape::half_type>;
 template struct test_gemm_transposea<migraphx::shape::fp8e4m3fnuz_type>;
+template struct test_gemm_transposea<migraphx::shape::fp8e5m2fnuz_type>;
 // TODO need hipblaslt support
 // template struct test_gemm_transposea<migraphx::shape::fp8e4m3fn_type>;
 // template struct test_gemm_transposea<migraphx::shape::fp8e5m2_type>;
```
Comment (TedThemistokleous) [inline test/verify/test_gemm_transposea.cpp:52]: Anyway to workaround and run these tests without hipblastlt?

File: test/verify/test_multinomial.cpp [hunk 5]
```diff
@@ -64,3 +64,6 @@ template struct test_multinomial<migraphx::shape::float_type>;
 template struct test_multinomial<migraphx::shape::half_type>;
 // TODO This fails, need to figure out why
```
Comment (TedThemistokleous) [inline test/verify/test_multinomial.cpp:65]: Is there an issue open for this or are we tracking this in any way?

**Review by TedThemistokleous — COMMENTED** [review]
Initial comments. Few questions, nothing concerning after you fixed the API compatability. 

Just more following up on some of your comments with tests and things.

File: test/verify/test_multinomial.cpp (hunk 5 above)
Comment (CharlieL7) [reply test/verify/test_multinomial.cpp:65]: yup, https://github.com/ROCm/AMDMIGraphX/issues/2516

File: src/include/migraphx/half.hpp [hunk 6]
```diff
@@ -80,6 +80,18 @@ struct common_type<migraphx::half, migraphx::fp8::fp8e4m3fnuz>
...
+struct common_type<migraphx::fp8::fp8e5m2fnuz, migraphx::half>
+{
+    using type = float;
+};
+
+template <>
+struct common_type<migraphx::half, migraphx::fp8::fp8e5m2fnuz>
+{
+    using type = float;
+};
```
Comment (pfultz2) [inline src/include/migraphx/half.hpp:93]: So #3606 refactors this to use the template type so these specializations are not necessary.

File: test/verify/test_gemm_add.cpp (hunk 3 above)
Comment (CharlieL7) [reply test/verify/test_gemm_add.cpp:63]: these don't work, umang made an issue about them earlier

**Review by pfultz2 — APPROVED** [review]

**Review by TedThemistokleous — APPROVED** [review]

**Comment by TedThemistokleous** [comment]
Should get this in sooner than later @causten . Onnxrt has support for this as well form the looks of it so it would be nice to get his all in for fp8 support at once

**Pull request merged**